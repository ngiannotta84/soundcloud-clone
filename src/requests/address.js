// const address = "https://soundcloud-clone-api.onrender.com";
const address = "http://localhost:4000";

export default address;
