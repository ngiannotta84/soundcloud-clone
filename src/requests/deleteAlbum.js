import { deleteRequest } from "./helpers";

const deleteAlbum = async (id) => {
  const response = await deleteRequest("albums", id);
  return response;
};

export default deleteAlbum;
