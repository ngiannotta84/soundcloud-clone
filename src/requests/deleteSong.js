import { deleteRequest } from "./helpers";

const deleteSong = async (id) => {
  try {
    const response = await deleteRequest("songs", id);
    return response;
  } catch (err) {
    throw new Error(err);
  }
};

export default deleteSong;
