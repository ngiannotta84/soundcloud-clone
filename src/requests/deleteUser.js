import { deleteRequest } from "./helpers";

const deleteUser = async (id, password) => {
  const endpoint = `${id}/${password}`;

  const response = await deleteRequest("users", endpoint);
  return response;
};

export default deleteUser;
