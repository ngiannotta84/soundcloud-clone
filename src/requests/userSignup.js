import Cookies from "js-cookie";
import { postRequest } from "./helpers";

const userSignup = async (data) => {
  try {
    const response = await postRequest("users/signup", data);
    Cookies.set("userToken", response.userToken, { expires: 1 });
    return response;
  } catch (err) {
    throw new Error(err);
  }
};

export default userSignup;
