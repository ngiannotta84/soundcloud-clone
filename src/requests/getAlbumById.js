/* eslint-disable no-console */
import axios from "axios";
import address from "./address";

const getAlbumById = async (id) => {
  try {
    const response = await axios.get(`${address}/albums/${id}`);
    return response.data;
  } catch (err) {
    console.error(err);
    throw new Error(err);
  }
};

export default getAlbumById;
