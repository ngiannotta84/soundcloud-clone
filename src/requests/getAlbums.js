/* eslint-disable no-console */
import axios from "axios";
import address from "./address";

const getAlbums = async ({ name, exact, limit }) => {
  let endpoint = `${address}/albums`;
  const queryArray = [];
  if (name) {
    queryArray.push(`name=${name}`);
    if (exact) {
      queryArray.push("exact=true");
    }
  }
  if (limit) {
    queryArray.push(`limit=${limit}`);
  }
  if (queryArray.length > 0) {
    const query = queryArray.join("&");
    endpoint = `${endpoint}?${query}`;
  }
  console.log(endpoint);
  try {
    const response = await axios.get(endpoint);
    console.log(response);
    return response.data;
  } catch (err) {
    console.error(err);
    throw new Error(err);
  }
};

export default getAlbums;
