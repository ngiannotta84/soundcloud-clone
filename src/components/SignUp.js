import { React, useState } from "react";
import PropTypes from "prop-types";
import { useNavigate, Link } from "react-router-dom";
import Alert from "./Alert";
import userSignup from "../requests/userSignup";

const SignUp = ({ handleLogin }) => {
  const initialState = {
    fields: {
      username: "",
      email: "",
      password: "",
      passwordConfirmation: "",
    },
    alert: "",
  };
  const [fields, setFields] = useState(initialState.fields);
  const [alert, setAlert] = useState(initialState.alert);
  const navigate = useNavigate();

  const handleFieldChange = (e) => {
    setFields({ ...fields, [e.target.name]: e.target.value });
  };

  const handleCredentials = async (event) => {
    const EMAIL_REGEX = /[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,4}$/;
    event.preventDefault();
    if (!fields.username) {
      setAlert("please provide a username");
    } else if (!fields.email) {
      setAlert("please provide your email address");
    } else if (!fields.email.match(EMAIL_REGEX)) {
      setAlert("please provide a valid email");
    } else if (!fields.password) {
      setAlert("please insert your password");
    } else if (fields.password.length < 8) {
      setAlert("password must be atleast 8 characters long");
    } else if (fields.passwordConfirmation !== fields.password) {
      setAlert("passwords must match");
    } else {
      try {
        const response = await userSignup({
          name: fields.username,
          email: fields.email,
          password: fields.password,
        });
        handleLogin(response);
        navigate(`/profile/${response.name}`);
      } catch (err) {
        setAlert(err.message);
      }
    }
  };

  return (
    <div className="login">
      <Alert message={alert} />
      <form onSubmit={handleCredentials} className="login__form">
        <h2 className="login__heading">Please Signup</h2>
        <label htmlFor="username" className="login__label">
          <span className="signup__label-text">Username</span>
          <input
            type="text"
            id="username"
            name="username"
            value={fields.username}
            onChange={handleFieldChange}
            className="login__input"
          />
        </label>
        <label htmlFor="email" className="login__label">
          <span className="signup__label-text">Email</span>
          <input
            type="text"
            id="email"
            name="email"
            value={fields.email}
            onChange={handleFieldChange}
            className="login__input"
          />
        </label>
        <label htmlFor="password" className="login__label">
          <span className="signup__label-text">Password</span>
          <input
            type="password"
            id="password"
            name="password"
            value={fields.password}
            onChange={handleFieldChange}
            className="login__input"
          />
        </label>
        <label htmlFor="passwordConfirmation" className="login__label">
          <span className="signup__label-text">Confirm Password</span>
          <input
            type="password"
            id="passwordConfirmation"
            name="passwordConfirmation"
            value={fields.password_confirmation}
            onChange={handleFieldChange}
            className="login__input"
          />
        </label>
        <button className="login__button" type="submit">
          SignUp
        </button>
      </form>
      <p className="login__signup">
        Already have an account?
        <Link to="/login" className="login__signup__link">
          Log-in Here
        </Link>
      </p>
    </div>
  );
};

SignUp.propTypes = {
  handleLogin: PropTypes.func.isRequired,
};

export default SignUp;
