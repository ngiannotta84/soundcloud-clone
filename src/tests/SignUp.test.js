import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import SignUp from "../components/SignUp";
import * as userSignup from "../requests/userSignup";

describe("Login", () => {
  const validProps = {
    handleLogin: jest.fn(),
  };
  const userSignupMock = jest.spyOn(userSignup, "default");

  test("snapshot", () => {
    const { asFragment } = render(
      <Router>
        <SignUp handleLogin={validProps.handleLogin} />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe("tests", () => {
    beforeEach(() => {
      render(
        <Router>
          <SignUp handleLogin={validProps.handleLogin} />
        </Router>
      );
    });

    test("renders correctly", () => {
      expect(screen.getByText(/please signup/i)).toBeInstanceOf(
        HTMLHeadingElement
      );
      expect(screen.getByLabelText(/username/i)).toBeInTheDocument();
      expect(screen.getByLabelText(/email/i)).toBeInTheDocument();
      expect(screen.getByLabelText(/^password$/i)).toBeInTheDocument();
      expect(screen.getByLabelText(/confirm password/i)).toBeInTheDocument();
      expect(screen.getByText(/submit/i)).toBeInstanceOf(HTMLButtonElement);
      expect(screen.getByText(/already have an account?/i)).toBeInTheDocument();
      expect(screen.getByText(/log in Here/i)).toBeInTheDocument();
    });

    test("it renders an error message if no username is entered", () => {
      const button = screen.getByText(/submit/i);
      fireEvent.click(button);

      expect(
        screen.getByText(/please provide a username/i)
      ).toBeInTheDocument();
    });

    test("it renders an error message if no email is entered", () => {
      const username = screen.getByLabelText(/username/i);
      const button = screen.getByText(/submit/i);

      fireEvent.change(username, { target: { value: "name" } });
      fireEvent.click(button);

      expect(
        screen.getByText(/please provide your email address here/i)
      ).toBeInTheDocument();
    });

    test("it renders an error message if no password is entered", () => {
      const username = screen.getByLabelText(/username/i);
      const email = screen.getByLabelText(/email/i);
      const button = screen.getByText(/submit/i);

      fireEvent.change(username, { target: { value: "name" } });
      fireEvent.change(email, { target: { value: "me@gmail.com" } });
      fireEvent.click(button);

      expect(
        screen.getByText(/please insert your password here/i)
      ).toBeInTheDocument();
    });

    test("it renders an error message if the password is less than 8 characters long", () => {
      const username = screen.getByLabelText(/username/i);
      const email = screen.getByLabelText(/email/i);
      const password = screen.getByLabelText(/^password$/i);
      const button = screen.getByText(/submit/i);

      fireEvent.change(username, { target: { value: "name" } });
      fireEvent.change(email, { target: { value: "megmail.com" } });
      fireEvent.change(password, { target: { value: "1234567" } });
      fireEvent.click(button);

      expect(
        screen.getByText(/password must be atleast 8 characters long/i)
      ).toBeInTheDocument();
    });

    test("it renders an error message if the passwords don't match", () => {
      const username = screen.getByLabelText(/username/i);
      const email = screen.getByLabelText(/email/i);
      const password = screen.getByLabelText(/^password$/i);
      const confirmPassword = screen.getByLabelText(/confirm password/i);
      const button = screen.getByText(/submit/i);

      fireEvent.change(username, { target: { value: "name" } });
      fireEvent.change(email, { target: { value: "megmail.com" } });
      fireEvent.change(password, { target: { value: "12345678" } });
      fireEvent.change(confirmPassword, { target: { value: "12345Nicola" } });
      fireEvent.click(button);

      expect(screen.getByText(/passwords must match/i)).toBeInTheDocument();
    });

    test("it renders an error message if incorrect type of email is entered", () => {
      const username = screen.getByLabelText(/username/i);
      const email = screen.getByLabelText(/email/i);
      const password = screen.getByLabelText(/^password$/i);
      const confirmPassword = screen.getByLabelText(/confirm password/i);
      const button = screen.getByText(/submit/i);

      fireEvent.change(username, { target: { value: "name" } });
      fireEvent.change(email, { target: { value: "megmail.com" } });
      fireEvent.change(password, { target: { value: "12345Nicola" } });
      fireEvent.change(confirmPassword, { target: { value: "12345Nicola" } });
      fireEvent.click(button);

      expect(
        screen.getByText(/please provide a valid email/i)
      ).toBeInTheDocument();
    });

    test("it renders and error message if userSignup throws an error", () => {
      const error = "Error!";
      userSignupMock.mockImplementation(() => {
        throw new Error(error);
      });
      const username = screen.getByLabelText(/username/i);
      const email = screen.getByLabelText(/email/i);
      const password = screen.getByLabelText(/^password$/i);
      const confirmPassword = screen.getByLabelText(/confirm password/i);
      const button = screen.getByText(/submit/i);

      fireEvent.change(username, { target: { value: "name" } });
      fireEvent.change(email, { target: { value: "valid@email.com" } });
      fireEvent.change(password, { target: { value: "12345Nicola" } });
      fireEvent.change(confirmPassword, { target: { value: "12345Nicola" } });
      fireEvent.click(button);

      expect(screen.getByText(error)).toBeInTheDocument();
    });

    test("user is logged-in if userSignup is succesfull", async () => {
      const fakeResolve = {
        name: "fakeName",
        id: "fakeId",
      };
      userSignupMock.mockResolvedValue(fakeResolve);

      const username = screen.getByLabelText(/username/i);
      const email = screen.getByLabelText(/email/i);
      const password = screen.getByLabelText(/^password$/i);
      const confirmPassword = screen.getByLabelText(/confirm password/i);
      const button = screen.getByText(/submit/i);

      fireEvent.change(username, { target: { value: "name" } });
      fireEvent.change(email, { target: { value: "valid@email.com" } });
      fireEvent.change(password, { target: { value: "12345Nicola" } });
      fireEvent.change(confirmPassword, { target: { value: "12345Nicola" } });
      fireEvent.click(button);

      await waitFor(() => {
        expect(validProps.handleLogin).toBeCalledWith(fakeResolve);
      });
    });
  });
});
