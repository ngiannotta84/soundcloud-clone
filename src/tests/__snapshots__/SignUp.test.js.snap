// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login snapshot 1`] = `
<DocumentFragment>
  <div
    class="login"
  >
    <h2>
      Please Signup
    </h2>
    <form>
      <div
        class="form-field"
      >
        <label
          for="username"
        >
          <span>
            username
          </span>
          <input
            id="username"
            name="username"
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        class="form-field"
      >
        <label
          for="email"
        >
          <span>
            email
          </span>
          <input
            id="email"
            name="email"
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        class="form-field"
      >
        <label
          for="password"
        >
          <span>
            Password
          </span>
          <input
            id="password"
            name="password"
            type="password"
            value=""
          />
        </label>
      </div>
      <div
        class="form-field"
      >
        <label
          for="passwordConfirmation"
        >
          <span>
            Confirm Password
          </span>
          <input
            id="passwordConfirmation"
            name="passwordConfirmation"
            type="password"
            value=""
          />
        </label>
      </div>
      <div>
        <button
          class="submit-button"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
    <p>
      Already have an account?
      <a
        class="navbar-links-item"
        href="/login"
      >
        Log in Here
      </a>
    </p>
  </div>
</DocumentFragment>
`;
